client/tx_test.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Spencer Kimball (spencer@cockroachlabs.com)
+
+package client
+
+import (
+	"database/sql"
+	"fmt"
+	"sync"
+	"testing"
+
+	// Needed for postgres driver test.
+	"github.com/cockroachdb/cockroach-go/testserver"
+)
+
+// TestExecuteTx verifies transaction retry using the classic
+// example of write skew in bank account balance transfers.
+func TestExecuteTx(t *testing.T) {
+	db, stop := testserver.NewDBForTest(t)
+	defer stop()
+
+	initStmt := `
+CREATE DATABASE d;
+CREATE TABLE d.t (acct INT PRIMARY KEY, balance INT);
+INSERT INTO d.t (acct, balance) VALUES (1, 100), (2, 100);
+`
+	if _, err := db.Exec(initStmt); err != nil {
+		t.Fatal(err)
+	}
+
+	type queryI interface {
+		Query(string, ...interface{}) (*sql.Rows, error)
+	}
+
+	getBalances := func(q queryI) (bal1, bal2 int, err error) {
+		var rows *sql.Rows
+		rows, err = q.Query(`SELECT balance FROM d.t WHERE acct IN (1, 2);`)
+		if err != nil {
+			return
+		}
+		for i, bal := range []*int{&bal1, &bal2} {
+			if !rows.Next() {
+				err = fmt.Errorf("expected two balances; got %d", i)
+				return
+			}
+			if err = rows.Scan(bal); err != nil {
+				return
+			}
+		}
+		return
+	}
+
+	runTxn := func(wg *sync.WaitGroup, iter *int) <-chan error {
+		errCh := make(chan error, 1)
+		go func() {
+			*iter = 0
+			errCh <- ExecuteTx(db, func(tx *sql.Tx) error {
+				*iter++
+				bal1, bal2, err := getBalances(tx)
+				if err != nil {
+					return err
+				}
+				// If this is the first iteration, wait for the other tx to also read.
+				if *iter == 1 {
+					wg.Done()
+					wg.Wait()
+				}
+				// Now, subtrace from one account and give to the other.
+				if bal1 > bal2 {
+					if _, err := tx.Exec("UPDATE d.t SET balance=balance-100 WHERE acct=1; UPDATE d.t SET balance=balance+100 WHERE acct=2"); err != nil {
```

Comment:
This line is excessively long. I think it would be more readable to use a backtick string and spread it out over multiple lines.